src/parachain/escrow.ts

```diff
@@ -0,0 +1,243 @@
+    }
+
+    async createLock<U extends GovernanceUnit>(
+        amount: MonetaryAmount<Currency<U>, U>,
+        unlockHeight: number
+    ): Promise<void> {
+        const tx = this.api.tx.escrow.createLock(amount.toString(amount.currency.rawBase), unlockHeight);
+        await this.sendLogged(tx, this.api.events.escrow.Deposit);
+    }
+
+    async withdraw(): Promise<void> {
+        const tx = this.api.tx.escrow.withdraw();
+        await this.sendLogged(tx, this.api.events.escrow.Withdraw);
+    }
+
+    async withdrawRewards(): Promise<void> {
+        const tx = this.api.tx.escrowAnnuity.withdrawRewards();
+        // EscrowAnnuity::WithdrawRewards doesn't seem to emit any event
+        await this.sendLogged(tx);
+    }
+
+    async increaseAmount<U extends GovernanceUnit>(amount: MonetaryAmount<Currency<U>, U>): Promise<void> {
+        const tx = this.api.tx.escrow.increaseAmount(amount.toString(amount.currency.rawBase));
+        await this.sendLogged(tx, this.api.events.escrow.Deposit);
+    }
+
+    async increaseUnlockHeight(unlockHeight: number): Promise<void> {
+        const tx = this.api.tx.escrow.increaseUnlockHeight(unlockHeight);
+        await this.sendLogged(tx, this.api.events.escrow.Deposit);
+    }
+
+    async votingBalance(
+        accountId: AccountId,
+        blockNumber?: number
+    ): Promise<MonetaryAmount<Currency<GovernanceUnit>, GovernanceUnit>> {
+        const height = blockNumber ? blockNumber : await this.systemAPI.getCurrentBlockNumber();
+        const userPointEpoch = await this.api.query.escrow.userPointEpoch(accountId);
+        const lastPoint = await this.api.query.escrow.userPointHistory(accountId, userPointEpoch);
+        const rawBalance = this.rawBalanceAt(parseEscrowPoint(lastPoint), height);
+
+        // `rawBalance.toString()` is used to convert the BN to a BigSource type
+        return newMonetaryAmount(
+            rawBalance.toString(),
+            toVoting(this.governanceCurrency)
+        );
+    }
+
+    private rawBalanceAt(escrowPoint: RWEscrowPoint, height: number): BN {
+    /*
+    Rust reference implementation:
+    
+    pub fn balance_at(who: &T::AccountId, height: Option<T::BlockNumber>) -> BalanceOf<T> {
+        let height = height.unwrap_or(Self::current_height());
+        let last_point = <UserPointHistory<T>>::get(who, <UserPointEpoch<T>>::get(who));
+        last_point.balance_at::<T::BlockNumberToBalance>(height)
+    }
+    */
+        const heightDiff = new BN(height).sub(escrowPoint.ts);
+        return escrowPoint.bias.sub(escrowPoint.slope.mul(heightDiff));
+    }
+
+    async totalVotingSupply(
+        blockNumber?: number
+    ): Promise<MonetaryAmount<Currency<VoteUnit>, VoteUnit>> {
+        const [currentBlockNumber, epoch, span, rawSlopeChanges] = await Promise.all([
+            this.systemAPI.getCurrentBlockNumber(),
+            this.api.query.escrow.epoch(),
+            this.getSpan(),
+            this.api.query.escrow.slopeChanges.entries()
```

Comment:
This may grow very big IIRC.

Reply:
In practice this wouldn't be called at all. What I did was to separate the data fetching from the logic so `index` can reuse the logic. Then, the UI would call index exclusively. I'll add a doc `@remark` about the performance of this getter though.